khmer/utils.py

```diff
@@ -18,9 +18,89 @@ def print_error(msg):
     print >>sys.stderr, msg
 
 
+def check_is_pair(record1, record2):
+    """
+    Checks to see if the two sequence records are left and right pairs
+    of each other, respectively.  Returns True or False as appropriate.
+
+    Handles both Casava formats: seq/1 and seq/2, and 'seq::... 1::...'
+    and 'seq::... 2::...'.
+    """
+    is_fastq = False
+    if hasattr(record1, 'accuracy') or hasattr(record2, 'accuracy'):
+        if not (hasattr(record1, 'accuracy') and hasattr(record2, 'accuracy')):
+            raise ValueError("both records must be same type (FASTA or FASTQ)")
+        is_fastq = True
+
+    name1 = record1.name
+    name2 = record2.name
+
+    if ' ' in name1:                        # handle '@name 1:rst'
+        name1, rest1 = record1.name.split(' ', 1)
+        name2, rest2 = record2.name.split(' ', 1)
+
+        if name1 == name2 and \
+           rest1.startswith('1:') and rest2.startswith('2:'):
+            return True
+
+    elif name1.endswith('/1') and name2.endswith('/2'):  # handle name/1
+        subpart1 = name1.split('/', 1)[0]
+        subpart2 = name2.split('/', 1)[0]
+
+        assert subpart1
+        if subpart1 == subpart2:
+            return True
+
+    return False
+
+
+def broken_paired_reader(screed_iter):
+    """
+    A generator that yields singletons and pairs from a stream of FASTA/FASTQ
+    records (yielded by 'screed_iter').  Yields (n, is_pair, r1, r2) where
+    'r2' is None if is_pair is False.
+
+    The input stream can be fully single-ended reads, interleaved paired-end
+    reads, or paired-end reads with orphans, a.k.a. "broken paired".
+
+    Usage::
+
+       for n, is_pair, read1, read2 in broken_paired_reader(...):
+          ...
+
+    Note that 'n' is the number of records read from the input stream, so
+    is incremented by 2 for a pair of reads.
+    """
+    record = None
+    prev_record = None
+
+    # handle the majority of the stream.
+    for n, record in enumerate(screed_iter):
+        if prev_record:
+            if check_is_pair(prev_record, record):
+                yield n, True, prev_record, record  # it's a pair!
```

Comment:
Why include the `n` as a return value at all? It's more common to let the user decide if they want to enumerate.